core/tls-websocket-server/src/lib.rs

```diff
@@ -36,87 +39,127 @@ pub mod sgx_reexport_prelude {
 #[cfg(all(not(feature = "std"), feature = "sgx"))]
 use crate::sgx_reexport_prelude::*;
 
-extern crate alloc;
-
-use crate::{connection::TungsteniteWsConnection, ws_server::TungsteniteWsServer};
-use alloc::boxed::Box;
-use log::*;
+use crate::{
+	config_provider::FromFileConfigProvider,
+	connection_id_generator::{ConnectionId, ConnectionIdGenerator},
+	error::WebSocketResult,
+	ws_server::TungsteniteWsServer,
+};
+use mio::{event::Evented, Token};
 use std::{
-	io::Error as IoError,
-	net::AddrParseError,
+	fmt::Debug,
 	string::{String, ToString},
 	sync::Arc,
 };
 
-mod common;
-pub mod connection;
-mod ws_server;
-
-/// General web-socket error type
-#[derive(Debug, thiserror::Error)]
-pub enum WebSocketError {
-	#[error("Invalid certificate error: {0}")]
-	InvalidCertificate(String),
-	#[error("Invalid private key error: {0}")]
-	InvalidPrivateKey(String),
-	#[error("Invalid web-socket address error: {0}")]
-	InvalidWsAddress(AddrParseError),
-	#[error("TCP bind error: {0}")]
-	TcpBindError(IoError),
-	#[error("Web-socket hand shake error")]
-	HandShakeError,
-	#[error("Web-socket connection already closed error")]
-	ConnectionClosed,
-	#[error("Web-socket connection has not yet been established")]
-	ConnectionNotYetEstablished,
-	#[error("Web-socket write error: {0}")]
-	SocketWriteError(String),
-	#[error("Web-socket handler error: {0}")]
-	HandlerError(Box<dyn std::error::Error + Sync + Send + 'static>),
-}
+pub mod config_provider;
+mod connection;
+pub mod connection_id_generator;
+pub mod error;
+mod tls_common;
+pub mod ws_server;
 
-pub type WebSocketResult<T> = Result<T, WebSocketError>;
+#[cfg(test)]
+mod test;
 
-/// abstraction of a web socket connection
-pub trait WebSocketConnection: Send + Sync {
-	fn process_request<F>(&mut self, initial_call: F) -> WebSocketResult<String>
-	where
-		F: Fn(&str) -> String;
+/// Connection token alias.
+#[derive(Eq, PartialEq, Clone, Copy, Debug)]
+pub struct ConnectionToken(pub usize);
```

Comment:
This type is like an alias to `mio::Token`. That way, we don't pollute any clients with `mio` Types.

Reply:
I was just wondering, what is the benefit here over re-exporting the `mio::Token`. Wouldn't this theoretically be better because a customer would be able to write its own client without needing to depend on us?

I have absolutely nothing against the decision because maybe we want this to be our own type to reduce potential mismatches if they have another mio version.

I don't really know about the best practices here. I am just wondering. :)

Reply:
If a client does not want to depend on us, it also does not depend on our `tls-websocket-server` crate? Or what kind of 'client' do you mean? What I meant in my comment is a 'client', or rather a dependent of our server crate, which is an additional server layer still (like our RPC server), not a client in the sense of a 'web-socket client'.
And for someone taking our websocket server crate and building another server layer on top of it, I find it better design to hide the `mio` dependency, since it is an implementation detail and does not need to be a concern of any user of this crate.

Reply:
Ok, thanks for your reasonable explanation. I think you are right. I have been reading up on this too meanwhile, and it seems that if possible you should not have external types part of your API, see [reddit](https://www.reddit.com/r/rust/comments/b7n5ue/is_it_ok_to_publicly_reexport_types_from_3rd/).

And since this is an implementation detail as you said, it should better be hidden.